src/coreclr/jit/lsra.cpp

```diff
@@ -8300,8 +8300,16 @@ void LinearScan::handleOutgoingCriticalEdges(BasicBlock* block)
     }
 
 #ifdef TARGET_ARM64
-    // Next, if this blocks ends with a JCMP, we have to make sure not to copy
-    // into the register that it uses or modify the local variable it must consume
+    // Next, if this blocks ends with a JCMP, we have to make sure:
+    // 1. Not to copy into the register that JCMP uses
+    //    e.g. JCMP w21, BRANCH
+    // 2. Not to copy into the source of JCMP's operand before it is consumed
+    //    e.g. Should not use w0 since it will contain wrong value after resolution
+    //          call METHOD
+    //          ; mov w0, w19  <-- should not resolve in w0 here.
+    //          mov w21, w0
+    //          JCMP w21, BRANCH
```

Comment:
Not obvious to me why the copy needs special casing here -- could we not insert the resolution move between the copy and jump?

Reply:
Inserting the resolution move between the copy and jump should be fine for some cases, for example:
```asm
mov     w21, w0
mov     x0, x19  ; <-- resolution move
cbnz    w21, G_M2478_IG37
```

but when we generate code for `JCMP`, it is tied with the copy and so the resolution move happens before the copy/jmp, like this:

```asm
      mov     x0, x20
      ldr     x21, [x20]
      ldr     x21, [x21,#72]
      ldr     x21, [x21,#40]
      blr     x21
      mov     x0, x19   ; move resolution overwrites `x0` before it gets copied.
      mov     w21, w0
      cbnz    w21, G_M2478_IG37

G_M2478_IG05:              
      mov     x19, x0
      ....
```

Reply:
Maybe add a comment that the `GT_COPY` gets special handling in codegen and is conceptually merged with the operation that consumes its result, so both its input and output regs must be excluded from the set available for resolution.

And only switches and jcmp have input regs (and so can be fed by copies), so those are the only block-ending branches that need special handling.

Which makes me wonder if we might need similar logic for switches. It looks like we don't check if the input regs for switches come from `GT_COPY`. Switches are rare enough that we might not have good stress coverage.

Reply:
> Maybe add a comment

Done.

> wonder if we might need similar logic for switches

Possibly, but I am not confident to make that change before seeing a scenario, so lets postpone it for now.